cpmpy/solvers/scip.py

```diff
@@ -0,0 +1,419 @@
+        cpm_cons = toplevel_list(cpm_expr)
+        supported = {"alldifferent"}  # alldiff has a specialized MIP decomp in linearize
+        cpm_cons = decompose_in_tree(cpm_cons, supported)
+        cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
+        cpm_cons = reify_rewrite(cpm_cons, supported=frozenset(['sum', 'wsum']))  # constraints that support reification
+        cpm_cons = only_numexpr_equality(cpm_cons, supported=frozenset(["sum", "wsum", "sub"]))  # supports >, <, !=
+        cpm_cons = only_bv_implies(cpm_cons)  # anything that can create full reif should go above...
+        cpm_cons = linearize_constraint(cpm_cons, supported=frozenset({"sum", "wsum","sub", "mul", "div"})) # the core of the MIP-linearization
+        cpm_cons = only_positive_bv(cpm_cons)  # after linearization, rewrite ~bv into 1-bv
+        return cpm_cons
+
+    def __add__(self, cpm_expr_orig):
+      """
+            Eagerly add a constraint to the underlying solver.
+
+            Any CPMpy expression given is immediately transformed (through `transform()`)
+            and then posted to the solver in this function.
+
+            This can raise 'NotImplementedError' for any constraint not supported after transformation
+
+            The variables used in expressions given to add are stored as 'user variables'. Those are the only ones
+            the user knows and cares about (and will be populated with a value after solve). All other variables
+            are auxiliary variables created by transformations.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: self
+      """
+      import pyscipopt as scip
+
+      # add new user vars to the set
+      get_variables(cpm_expr_orig, collect=self.user_vars)
+
+      # transform and post the constraints
+      for cpm_expr in self.transform(cpm_expr_orig):
+
+        # Comparisons: only numeric ones as 'only_bv_implies()' has removed the '==' reification for Boolean expressions
+        # numexpr `comp` bvar|const
+        if isinstance(cpm_expr, Comparison):
+            lhs, rhs = cpm_expr.args
+            sciprhs = self.solver_var(rhs)
+
+            # Thanks to `only_numexpr_equality()` only supported comparisons should remain
+            if cpm_expr.name == '<=':
+                if (isinstance(lhs, Operator) and lhs.name == "sum" and all(a.is_bool() and not isinstance(a, NegBoolView) for a in lhs.args)):
+                    if rhs == 1: # special SOS1 constraint?
+                        self.scip_model.addConsSOS1(self.solver_vars(lhs.args))
+                    else: # cardinality constraint
+                        self.scip_model.addConsCardinality(self.solver_vars(lhs.args), rhs)
+                else:
+                    sciplhs = self._make_numexpr(lhs)
+                    self.scip_model.addCons(sciplhs <= sciprhs)
+
+            elif cpm_expr.name == '>=':
```

Comment:
Just a general interest: Due to how the constraints are transformed are you certain that these consraints can't be multiplied by -1 to become cardinality constraints?

Reply:
Aha no, actually we are not... I will do a check for it here

Reply:
Aha actually no... I'll add a check for it here